skfda/misc/_math.py

```diff
@@ -679,3 +679,38 @@ def cosine_similarity_matrix(
     return _clip_cosine(
         inner_matrix / norm1[:, np.newaxis] / norm2[np.newaxis, :],
     )
+
+def half_sq_euclidean(
+    arg1: NDArrayFloat,
+    arg2: Optional[NDArrayFloat] = None,
+) -> NDArrayFloat:
+    """
+    Return the half squared Euclidean row distance between
+    two matrices.
+    
+    Args:
+        arg1: First sample (n_samples_1, n_features).
+        arg2: Second sample (n_samples_2, n_features).
+            If it is ``None`` returns the half squared Euclidean distance
+            between the rows of ``arg1``.
+
+    The distance is taken w.r.t matrix rows.
+    """
+    if arg2 is not None:
+        # np.dot(X, Y.T)=inner_product_matrix(X, Y) but np ~1.6 faster
```

Comment:
Yes. I profiled it with `%timeit`.
`np.dot(X, Y.T)` are `np.einsum('n...,m...->nm...', arg1, arg2).sum(axis=-1)` (see line 327) are equivalent: it is an outer product. I don't use the `np.einsum` and I guess there may have been a reason of using it instead `np.dot`(??)
So at least line 327, I'd replace the np.einsum by np.dot

Reply:
And same story for `np.sum(arg1 ** 2, axis=1)` (would be `np.diag(inner_product_matrix(arg1, arg1))`

Reply:
Note that both `inner_product` and `inner_product_matrix` take into account the spacing between grid points (which may be uneven). I am not sure if that is necessary in this algorithm, but it is different than just the sum. Also `np.diag(inner_product_matrix(arg1, arg1))` is slow because you compute all values of the matrix. It is better to do `inner_product(arg1, arg1))`.

Reply:
Yes, it is not necessary for this algorithm because we explicitly compute the distances between the sampled version of the two functions.
I've put  ìnner_product`

Reply:
Wait, I forgot that this code uses `inner_product` between vectors and not functions, so yeah, they are no grid points and the two are equivalent.